### ISSUE TITLE
[TW] Use LUKS2 and offer TPM-auth in installer

### ISSUE THREAD

**Issue by bbrunner** [yast/yast-installation#1088]
While the YaST-installer in Tumbleweed still uses LUKS1 for Full Disk Encryption as default (to enable LUKS2 you have to add "YAST_LUKS2_AVAILABLE=1" as kernel command line option), Agama on ALP already supports LUKS2 with the additional authentication mechanism via TPM2.0.

With the last SRs (sr#1090915 and sr#1090821) being accepted into Factory, the sources in ALP and Factory are finally in sync and IMHO it would be a good time to do the switch from LUKSv1 to v2 with the same defaults as we have in ALP.

While grub2 currently does not support Argon2, we are currently tied to use PBKDF2 as encryption-algorithms. Nevertheless, there are also plans to introduce Argon2 support.
Therefore, it would be nice if the functionality could be easily extended to select an Algorithm in the future.

Please let me know if you need any additional information.

**Comment by bbrunner** [comment]
Are there any news, or do you need any additional information?

**Comment by bbrunner** [comment]
Thanks Ancor for the update!
In regards to LUKS2 (without TPM), it should already be safe to use, if the selected default-algorythm for the password is PBKDF2 instead of Argon2.

But after the TPM enablement is already in progress, it should be fine to wait till everything is in place, before moving it to the "regular" installer.

**Comment by ancorgs** [comment]
> [...] In regards to LUKS2 (without TPM), it should already be safe to use, if the selected default-algorythm for the password is PBKDF2 instead of Argon2.

Yes. But that could give users a false sense of security. Since LUKS2 with PBKDF2 is not really more secure than LUKS1.

**Comment by lnussel** [comment]
any news on this one? With systemd-boot, encryption can safely use luks2 with argon

**Comment by lnussel** [comment]
Given that luks2 metadata is precondition for many other features systemd brings such as TPM, FIDO2 and recovery keys, and grub2 argon support is still not there even after more than six months, maybe we need to reconsider. I'm not a huge fan of a non-default luks2 setup ie using it with pbkdf2 but luks1 is really getting in our way now.

**Comment by aplanas** [comment]
> LUKS2 with PBKDF2 is not really more secure than LUKS1

That is true but defaulting to LUKS2 always will help us to introduce the FDE based on systemd that we have for MicroOS / TW. The selection of using the default Argon2 or downgrade to PBKDF2 can be done later, when the user select systemd or grub based fde.

In any case LUKS2 is an invariant.

**Comment by ancorgs** [comment]
In Agama we are already going for LUKS2 with PBKDF2 (with TPM unlocking if possible and chosen by the user). See https://github.com/openSUSE/agama/pull/995 (still not merged).

I guess the same (forcing PBKDF2) is acceptable for TW installed with YaST.

**Comment by lnussel** [comment]
So what's needed to move this forward?

**Comment by ancorgs** [comment]
Ok. Time for next steps. Sorry for not being responsive, but I was busy with other stuff (Agama-related, mostly).

Bringing TPM-auth to YaST would be quite some work. I will summarize below everything that is missing in case somebody wants to try.

But unleashing LUKS2 in YaST for TW should not be much of a problem, so I think it's time to take the next steps.

### Introducing LUKS2 support

- First, I would remove the need to use `YAST_LUKS2_AVAILABLE` by removing [this method definition](https://github.com/yast/yast-storage-ng/blob/123b26f49b0ecbebc015b63afafb9716f8193938/src/lib/y2storage/encryption_method/luks2.rb#L64)
  - That would make LUKS2 available to anyone using the Expert Partitioner.
    - All my initial concerns about people using this wrongly and then opening bug reports still stand (I expect unpredictable out-of-memory errors and users wrongly using Argon2 for file-systems needed for booting).
    - We may mitigate that by setting PBKDF2 as the default value in the Expert Partitioner interface.
  - That would not change the default encryption technology used by the Guided Setup. It would still be LUKS1.
- Then I would make the Guided Setup configurable by product/role in the control file
  - That implies adding the fields `ProposalSettings#encryption_method` and `ProposalSettings#encryption_pbkdf` to `#load_features`
  - It would also be good to document that [here](https://github.com/yast/yast-installation/blob/master/doc/control-file.md#the-proposal-subsection).

**Comment by lnussel** [comment]
Could we just make luks2 with pbkdf the default?

Roles are already used for desktop selection IIRC. So unless there are layers of roles we'd have to duplicate all of them to offer FDE.

**Comment by ancorgs** [comment]
> Could we just make luks2 with pbkdf the default?

I guess we could. But I would also do this part then:

> We may mitigate that by setting PBKDF2 as the default value in the Expert Partitioner interface.

Right now, the default pbkdf2 is a setting of the storage proposal (a.k.a. Guided Setup). I would make the Expert Partitioner also honor that setting.

**Comment by lnussel** [comment]
Ok, sure. If there isn't one central place that sets the default kdf everywhere then lets change all places.

**Comment by schubi2** [comment]
>>Could we just make luks2 with pbkdf the default?
>I guess we could.
But then when and where should be the password input by the user ?

**Comment by ancorgs** [comment]
My own summary :-)

- In Tumbleweed, we should use LUKS2 with PBKDF2 as default encryption settings for anyone using encryption (no matter whether they use the default Grub2 or systemd-boot, whether they use the Expert Partitioner or Guided Setup, etc.). That implies
  - Change the value of `ProposalSettings#encryption_method` and `ProposalSettings#encryption_pbkdf` for Tumbleweed.
    - I would suggest to do that by adding those fields to `ProposalSettings#load_features` and then modifying the control file for Tumbleweed.
    - Ludwig suggested me in a call a more hacky approach of simply modifying the default values with a patch in the spec file that would only be applied at TW.
  - Make LUKS2 [always available](https://github.com/yast/yast-storage-ng/blob/123b26f49b0ecbebc015b63afafb9716f8193938/src/lib/y2storage/encryption_method/luks2.rb#L64) in the Expert Partitioner.
  - Always use LUKS2 + PBKDF2 as default value in the Expert Partitioner form for encryption.
    - I would suggest to do that by honoring `ProposalSettings#encryption_method` [here](https://github.com/yast/yast-storage-ng/blob/123b26f49b0ecbebc015b63afafb9716f8193938/src/lib/y2partitioner/actions/controllers/encryption.rb#L214) and `ProposalSettings#encryption_pbkdf` [here](https://github.com/yast/yast-storage-ng/blob/123b26f49b0ecbebc015b63afafb9716f8193938/src/lib/y2partitioner/actions/controllers/encryption.rb#L74)
    - You can also follow Ludwig's hacky approach and patch that file at the spec file for TW.

- I still don't see the need of having an extra form or field or workflow step or anything for enabling encryption and for asking the password. With the above changes, anyone could just use the Guided Setup or the Expert Partitioner to enable encryption in the usual way and all Tumbleweed users would get a LUKS2+PBKDF2 system.

- About opting for systemd-boot or Grub2, I don't have an opinion about what is better (expert console or just a new setting at bootloader UI or just forcing it for TW...).

**Comment by aplanas** [comment]
> My own summary :-)
> 
>     * In Tumbleweed, we should use LUKS2 with PBKDF2 as default encryption settings for anyone using encryption (no matter whether they use the default Grub2 or systemd-boot, whether they use the Expert Partitioner or Guided Setup, etc.). That implies

I fail to see the logic here. There are two kinds of FDE as now:

* grub-base encryption: LUKS2 + PBKDF2
* user-space encryption: LUKS2 + argon2id (cryptsetup default)

The user space encryption is agnostic of the bootloader: you can use grub2 (+bls patch) or systemd-boot.

For user safety seems more logical to default to LUKS2 + argon2id unless the grub-base encryption is selected. Or what am I missing?

**Comment by lnussel** [comment]
The idea would be to start with the most simple and least intrusive change to get this issue moving at all finally. So first switch to the smallest common denominator of both setups, ie luks2+pbkdf2. Then we can improve the logic from there.
